docs/design_notes/module_lifecycle.md

```diff
@@ -0,0 +1,289 @@
+
+```
+
+
+A compact `Module` is similar in spirit to a function. It offers a concise notation and restricts external interaction to the inputs and return values of the function. In this case the concise notation might make it easier for others to understand what the Module is doing. There is no need to jump back and forth between the setup and call function to understand what the submodules are doing. Instead, simply reading the call method from top to bottom once should give a clear picture of what the Module is doing. This can make a significant difference if you are implementing complex Modules with many hyper parameters.
+
+Another benefit of defining submodules and/or variables inline is that you can use arguments to your method when constructing variables. The most common example of this is using shape information to determine the shape of a parameter like this:
+
+
+```python
+class CompactScaledMLP(nn.Module):
+  hidden_size: int
+  out_size: int
+
+  @nn.compact
+  def __call__(self, x):
+    scale = self.param("scale", x.shape[-1:], nn.initializers.ones)
+    x *= scale
+    a = nn.Dense(self.hidden_size)(x)
+    h = nn.relu(a)
+    return nn.Dense(self.out_size)(h)
+```
+
+
+Many of the standard linen layers like `nn.Dense` use this type of shape inference already to avoid the need to specify input shapes (like the number of input features to a Dense layer).
+
+### Compact control flow
+
+The order in which you define submodules determines the name of a submodule if none is provided explicitly (using the name= construction argument). Because the name determines how parameters are mapped to submodules, you must be careful about mixing control flow with auto-generated names. Using control flow can change the order or remove certain submodules altogether. This is useful in the case where certain submodules should only exist depending on some construction argument. However, when control flow depends on the input arguments to the Module you should be careful. For example the following Module will break:
+
+
+```python
+class WrongModule(nn.Module):
+  @nn.compact
+  def __call__(self, x, mode):
+    if mode == "encode"
+      return nn.Dense(8)(x)
+    elif mode == "decode":
+      return nn.Dense(4)(x)
+```
+
+
+This problem can be solved in various ways: Provide explicit names, create the modules in setup, or move the constructor out of the control flow. The latter is done as follows:
+
+
+```python
+class CorrectModule(nn.Module):
+  @nn.compact
+  def __call__(self, x, mode):
+    encoder = nn.Dense(8)
+    decoder = nn.Dense(4)
+    if mode == "encode"
+      return encoder(x)
+    elif mode == "decode":
+      return decoder(x)
+```
+
+
+In the above example the construction order is fixed. After construction the submodules can be used in an arbitrary order.
+
+> Note \
+compact modules show a strong resemblance to [React hooks](https://reactjs.org/docs/hooks-custom.html)
+
+
+## Top-level Modules
```

Comment:
Should the main title here be "top-level modules"? Or rather something along the lines "Bound/unbound modules" ?

I think about "Flax code" (i.e. code inside modules) to always run in a "bound" state. It cannot be called "from normal Python code" because the top-level modules are in an "unbound" state. So one needs to call the bridging init/apply functions that will call the code inside the module in a "bound" state.

It's an interesting fact that "top-level modules" can never be in a bound state, but rather are cloned on-the-fly and the original copy (instance) remains unbound all the time ... But it feels not like the most important part of this story

Reply:
> I think about "Flax code" (i.e. code inside modules) to always run in a "bound" state.

This inside / outside Module code is exactly the intuition I'm trying to cultivate here. bound/unbound is the jargon top-level vs inside Module is the practical distinction. I would prefer to keep it intuitive.

Reply:
ok then I think I need to update my thinking here :)